[10.0] Reworked components registry

guewen opened the pull request:
At the creation of the connector (OpenERP 7.0), many architecture choices have been made
based on the old API of Odoo. One of my main concern was to get rid of the `cr,
uid, context` in the classes of the connector.  Now, two things have changed:

* Odoo is entirely using its new API
* I have years of look back on what works well or not, plus various
  implementation of connectors in several domains, plus many
  community/colleagues feedbacks.

That brought me to think more and more about the future of the connector and
how to correct the former design mistakes, and at the same time making it
easier to grasp.

The connector is originally composed of:

* A Jobs Queue
* Events (producer, subscriber) pattern
* The Backends registries (`ConnectorUnit`) that I will now call the Components
  registry and a set of base components (`Mapper`, `Binder`, ...)

## Queue Job API

I already started with the Jobs Queue which is now extracted in an [independent
addon](https://github.com/OCA/queue/tree/10.0/queue_job) and has got its API
entirely reviewed. It is now melted into the Odoo API:

```python

class MyModel(models.Model):
   _name = 'my.model'

   @api.multi
   @job
   def my_method(self, a, k=None):
       _logger.info('executed with a: %s and k: %s', a, k)


class MyOtherModel(models.Model):
    _name = 'my.other.model'

    @api.multi
    def button_do_stuff(self):
        self.env['my.model'].with_delay().my_method('a', k=2)

```

## Components API

The basic idea is to keep and improve what works and dump what doesn't (easy
isn't it?), I'll try to make a better component system, more close to the
Odoo API.

What works:

* The component system. It allows to put together the different pieces and
  that's really cool to be able to decouple them.  Being able to "ask" for a class
  without knowing it is powerful.
* The `ConnectorEnvironment` instance which is passed to each new
  `ConnectorUnit` transversally, so we can store info inside and keep it all
  along a synchronization flow.

What is flawed:

* The worst is the extensibility: it's easy to override something in a new
  backend version, but hard to create extension modules to improve an existing
  backend version.
* The backend versions: too hard to use.  Differences between versions of a
  backend reveals to be too rare to justify this feature. Furthermore, this
  is the main reason for the flaw number 1 so let's drop it.
* The lookup of the `ConnectorUnit` by class. It limits some usage as we need
  to know and import the class.  It could be more dynamic and flexible with
  strings.
* Only able to retrieve one component at a time.
* `model_name` was mandatory but in some case we want components without model,
  we had to use fake model names
* If we can reduce the number of imports along the way...

I started a `component` addon, which will be a dependency of the `connector`,
but will also be an independent and agnostic component registry integrated
with Odoo.

The components borrows the `_name` and `_inherit` of Odoo for a good
extensibility, and introduce a `_usage` attribute for the lookup (kind of
informal interface, but components with different `_usage` can share the same
interface). Examples of `_usage` in the know world could be `import.mapper` or
`backend.adapter`.

Components are registered towards a `_collection` which is the name of a Model
acting the same way the `Backend` today and they are applied on models with
`_apply_on`.

So `_name` and `_inherit` are used for extensibility, `_usage` and `_apply_on`
for finding the Components.

When we want to work with the Components, we instanciate a `WorkContext`
(replacement of `ConnectorEnvironment`) with the record of the collection
(backend record), this work context will be passed to every component used
during this work session.

Example:

```python


class ShopBackend(models.Model):

    _name = 'shop.backend'
    _inherit = ['collection.base']

    name = fields.Char()
    url = fields.Char()


class ShopBaseComponent(AbstractComponent):
    _ name = 'shop.base'
    _collection = 'shop.backend'


class SaleImporter(Component):
    _name = 'importer.shop.sale'
    _inherit = ['shop.base', 'importer']  # importer is a base component defined in 'connector'
    _usage = 'importer'  # used for retrieval of the component, in fact, it is already set in the 'importer' Component so we would not need to repeat it here
    _apply_on = ['shop.sale.order']

    def run(self, external_id):
        # ...
        sale_mapper = self.component(
            usage='import.mapper',
            # no need to specify the model we are working with
        )
        partner_mapper = self.component(
            usage='import.mapper',
            model_name='shop.res.partner',  # for a different model
        )
        # ...
        return ...


class SaleImportMapper(Component):
    _name = 'import.mapper.shop.sale'
    _inherit = ['shop.base', 'import.mapper']  # import.mapper is a base component defined in 'connector'

    _usage = 'import.mapper'  # used for retrieval of the component
    _apply_on = ['shop.sale.order']

    def map_record(self, record):
        return ...


class PartnerImportMapper(Component):
    _name = 'import.mapper.shop.partner'
    _inherit = ['shop.base', 'import.mapper']  # 'import.mapper' is a base component defined in 'connector'

    _usage = 'import.mapper'  # used for retrieval of the component
    _apply_on = ['shop.res.partner']

    def map_record(self, record):
        return ...


class SaleOrder(models.Model):

    _inherit = 'shop.sale.order'

    @api.multi
    def import_from_shop(self, external_id):
        backend = self.backend_id
        with backend.work_on('shop.sale.order') as work:
            importer = work.component(usage='importer')
            importer.run(external_id)



```

Also, the new system will allow to search and run all components for a usage.
In the following example, we run a main mapper, then run all mappers with
`_usage` `import.mapper.extras`, meaning that an addon can plug its own mapper
easily (but he might as well use `_inherit` to extend the original one):

```python

  values = self.component(usage='import.mapper').map_record(record).values()
  for mapper in self.many_components(usage='import.mapper.extras'):
      values.update(mapper.map_record(record).values())

```

It seems this design would correct the flaws I mentioned above.

The existing set of base components (`Binder`, `Mapper`, ...) have been
reimplemented in `connector` on top of the Components.

## Events API

Events are now in a new addon `component_event` that depend on `component`. The API got an upgrade too:

```python
# Notification

    class Base(models.AbstractModel):
        _inherit = 'base'

        @api.model
        def create(self, vals):
            record = super(Base, self).create(vals)
            self._event('on_record_create').notify(record, fields=vals.keys())
            return record

# Listeners:

    class MyEventListener(Component):
        _name = 'my.event.listener'
        _inherit = 'base.event.listener'

        def on_record_create(self, record, fields=None):
            _logger.info("%r has been created", record)

```

Commit:
Start prototype of a component system

The inheritance system is copied from the Odoo on, because it works well
for extending addons and because it will be easier to use for developers
accustomed to it.

Commit:
Start to implement the collection system

Commit:
Improve on the collections, work, ...

Commit:
Extract 'components' into an individual addon

Commit:
Add test_component

Commit:
Separate lookup by name and by usage

The favorite lookup should be by 'usage' (kinda interface) and by model
name ('components()' method).  The lookup by component name should
normally not be used as it reduces the flexibility.  Using a different
method for this lookup discourage its usage.  Also, the lookup by
component name completely ignores the current collection, which is bad
or not bad depending of what you want to achieve.

guewen commented:
As the PR  #211 for the migration of the connector  is not merged yet, we see too much commits.
What is worth to see is in `component` and perhaps `test_component`.

pedrobaeza commented:
@sergio-teruel, can you take a look at the proposal?

lmignon commented:
@guewen Do you reinvent `zope.component` https://pypi.python.org/pypi/zope.component?

guewen commented:
@lmignon I had a look at zope.component and there's certainly some inspiration. However, my goal is to have something very close to the Odoo API, reusing its extension mechanism (`_inherit`), and much more simple than zope.component.

lmignon commented:
@guewen You are right to avoid the use of `zope.component`. (As a former user/developer of plone I prefer to to no more use this lib that makes your code unreadable.. IMPOV)

Commit:
Start to use 'components' in connector

guewen commented:
This is a very very early work but here is what would change on an implementation (untested, only small parts migrated): 
connector-ecommerce: https://github.com/guewen/connector-ecommerce/commit/82bacb7233b6412a196dca253264dec62fead98b
magento connector: https://github.com/OCA/connector-magento/commit/3986233faffa45ca9a43a943f74c7b90efa1c33e

Commit:
Add core components in connector

Commit:
Various fixes in component

Commit:
Add AbstractComponent

Commit:
Remove overrides of attributes

when inheriting from multiple components, we might inadvertly override a
value with None

Commit:
Allow to share some components across collections

Used for instance by the 'ecommerce' components

Commit:
Replace unit_for by components

Commit:
Instanciate the component returned by name

Commit:
Get rid of MetaMapper

The mapping methods are now built by the component initialization when
the "aggregated" class is built.

Commit:
Make unwrap_binding compatible with components

Commit:
Add check to help find duplicate components

And remove automatic naming from class name, not explicit enough,
especially when we have to inherit from one.

We could take the total opposite and only use class names components:

class MyComponent(Component):
    ...

class MyComponentExtended(Component):
    _inherit = 'MyComponent'

But it would be less close to the odoo's API.

Commit:
Rename file, it confuses pytest

That thinks the file contains tests.

Commit:
Run tests after install

Because they need the registry to be initialized, otherwise no component
exist

Commit:
Fix pep8

Commit:
Add tests to component

============================================================================================================= test session starts =============================================================================================================
platform linux2 -- Python 2.7.9, pytest-3.0.7, py-1.4.33, pluggy-0.4.0
run-last-failure: run all (no recorded failures)
rootdir: /opt/odoo, inifile:
plugins: cov-2.5.1, odoo-0.2.2
collected 25 items
odoo/external-src/connector/component/tests/test_build_component.py .........
odoo/external-src/connector/component/tests/test_component.py ........
odoo/external-src/connector/component/tests/test_lookup.py ......
odoo/external-src/connector/component/tests/test_work_on.py ..
---------- coverage: platform linux2, python 2.7.9-final-0 -----------
Name                                                                  Stmts   Miss  Cover
-----------------------------------------------------------------------------------------
odoo/external-src/connector/component/__init__.py                         4      0   100%
odoo/external-src/connector/component/__manifest__.py                     1      1     0%
odoo/external-src/connector/component/base_components.py                  3      0   100%
odoo/external-src/connector/component/builder.py                         18      0   100%
odoo/external-src/connector/component/core.py                           155      6    96%
odoo/external-src/connector/component/exception.py                        3      0   100%
odoo/external-src/connector/component/models/__init__.py                  1      0   100%
odoo/external-src/connector/component/models/collection.py                8      0   100%
odoo/external-src/connector/component/tests/__init__.py                   4      0   100%
odoo/external-src/connector/component/tests/common.py                    26      2    92%
odoo/external-src/connector/component/tests/test_build_component.py     102      0   100%
odoo/external-src/connector/component/tests/test_component.py            56      0   100%
odoo/external-src/connector/component/tests/test_lookup.py               84      0   100%
odoo/external-src/connector/component/tests/test_work_on.py              26      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                                   491      9    98%
========================================================================================================== 25 passed in 0.76 seconds ==========================================================================================================

Commit:
Use 2 different methods for single/many lookup

the 'components' method had 2 different return types depending of the
'multi' argument.

Now we have 'component' or 'many_components' that return a Component
instance or a list of Component instances.

guewen commented:
`component` is now fully tested (98%). Still lacking a bit of documentation.

`connector` will (should) be backward compatible, ConnectorUnit or Component classes will both be usable.
I plan to remove the ConnectorUnit / Backend versions features in Odoo 11.0.
I have to update the docs and add tests for the new Components. Some duplication might be removed between ConnectorUnit and Component classes (in mapper notably), however I don't want to spend too much time on that as the former will be dropped.

The Magento Connector implementation is going well, the main module is fully functional, I have to rewrite the tests (using vcr.py) and update the doc. https://github.com/guewen/connector-magento/tree/10.0-components

Commit:
Document component

guewen commented:
`component` got a round of documentation.

Commit:
Add tests for connector mapper

Commit:
Add deprecation comments

Commit:
Remove some Mapper duplication

We keep the ComponentUnit classes though, there is some duplication
of code but it will be dropped in Odoo 11.0

Commit:
Add/move tests for advisory locks

Commit:
Fix error messages

Commit:
Update test_connector

Commit:
Update documentation

Commit:
Start to write a migration guide

Commit:
Draft new component_event addon

Proposing a new API based on components for the events

Commit:
Hold a component registry per database

Because they can have different addons

Commit:
Add a cache on the components lookups

Commit:
Add cache in component_event

So we don't need to iterate on the listeners to find the events each
time the event is triggered.

* rename producer -> collecter, which is more accurate regarding what it does
* rename the trigger function 'fire' -> 'notify', just a matter of
  preference here :)
* add a CollectedEvents class so we don't modify the state of the
collecter and we don't mix responsibilities

Commit:
Improve components classes

* Move the methods to get components from the components to the
WorkContext, it should really be its responsibilities to get them.
There are shorcuts on the components though.
My first implementation was like that, then I moved the lookup methods
on the components thinking that it would allow to customize them. But if
we did that, we would have a different behavior from component to
component which is bad, so if one really wanted to do that, better have
to work with a different subclass of WorkContext, which would have its
own behavior, consistently throughout all the work session.

* Allow to switch the collection using work_on(), not only the
  model_name

Commit:
Complete component_event

* Add documentation
* Add a few more test cases
* Prevent using component lookup methods from an EventWorkContext when
  there is no collection, but allow to switch to a WorkContext with
  collection
* Increase cache size (should be measured at some point...)
* Fix a few issues with collection being an empty model being False-ish,
  so added comparisons with None
* Add abstract Component with the base events (on_record_create,
  on_record_write, on_record_unlink)

Commit:
Move test_producer in deprecated

We'll now use the component_event events

Commit:
Add cachetools as requirements for Travis

Needed by 'component' and 'component_event'

Commit:
Prevent trigger of events before registry is ready

We return an empty list of events so nothing is triggered

Commit:
Continue the migration guide

Commit:
Rewrite some parts of the doc, notably code overview

guewen commented:
[conn-doc.tar.gz](https://github.com/OCA/connector/files/1085402/conn-doc.tar.gz)

Here's a local build of the docs, including a migration guide. /cc @pedrobaeza @sergio-teruel

Commit:
Fix test: post_install

Tests using odoo transactions must run post install, because during the
install the registry is not ready, so the components aren't neither.

guewen commented:
Docs uploaded on http://connector-rework-doc.surge.sh/
Here is the migration guide: http://connector-rework-doc.surge.sh/guides/migration_guide.html#migration-guide

Note: the url is temporary, should be on odoo-connector.com once merged

Commit:
Move base component in a components directory

For consistency, this is where components should go (as for models,
views, ...)

Commit:
Check that component registry is ready for events

Checking that the odoo registry is ready is not working:
in tests, the events are not working because they are run
just before the odoo registry is set to ready.

Commit:
Ignore W0104 pylint warnings in tests

Commit:
Remove doc about creating abstract listener

I'm not sure it is good actually. If you inherit from it for only one of
the method, the other methods will listen be executed for doing nothing,
but that's still a useless call of method...

guewen commented:
Travis pass, but runbot fails, the instance is up and running though.

guewen commented:
Unrelated but BTW who set the "Squash and merge" to be the default option for merge? I don't want it to be the default for this project. Can I have back my access to the admin settings on this project? (I can't even change the default branch and so on)

lmignon commented:
> Unrelated but BTW who set the "Squash and merge" to be the default option for merge?

@guewen For sure it's not me :smirk:

pedrobaeza commented:
@guewen AFAIK, this is recognized from your last choice merging something. It's not a project setting, so don't blame me :stuck_out_tongue_winking_eye:

guewen commented:
> @guewen AFAIK, this is recognized from your last choice merging something. It's not a project setting, so don't blame me

I'm not! 

Yes you right, the only setting is if we want to allow the merge/squash/rebase or not. I would still be glad to have more rights on the project :) But at least someone setting the 10.0 branch as default :)

sbidoul commented:
@guewen I gave Admin access to the Connectors Maintainer teams (@lmignon, @lasley and yourself so that should be safe). 

Fantastic work here, btw!

Commit:
Use self.work.env, in case self.collection is empty

(might be empty for EventWorkContext)

guewen commented:
@sbidoul thanks!

Commit:
We have now the 3 parts with a new API :)

lmignon commented on component_event/models/base.py:
```diff
@@ -0,0 +1,112 @@
+    It also notifies the following events:
+
+    * ``on_record_create``
+    * ``on_record_multi``
```
@guewen s/on_record_multi/on_record_write

lmignon commented:
@guewen Your PR is amazing! It's a long reading to understand all the changes and the benefits of this great refactoring. I'll try to use this new version in the next weeks on my current project to validate this new architecture. Thank you for this great work!

Commit:
Correct event name in doc and add signature

Commit:
Add Manager group on partner view connector's tab

Commit:
Remove RecordsEventListener

It's not really usable: if we inherit from it and implement only one
method, the others will be called without effect, waste of resources

Commit:
Add @skip_if decorator to skip events

It takes a condition which skip the event when evaluated to True.
A new base listener for the connector adds a default condition based on
self.env.context.get('connector_no_export'). Every export event listener
should be decorated with:

    @skip_if(lambda self, *args, **kwargs: self.no_connector_export)

Commit:
Change Collection.work_on() to a context manager

It allows to open/close objects on start/end of a synchro / work
session. The base method does none of that, but this is a really common
use case. Example: at the beginning of the work session, I open a
webservice client that is available in our WorkContext for the duration
of our work. At the end, I close the client to end the connection.

guewen commented:
@here if anyone is currently trying this branch, you should be aware of the change of a8e1778:

> Change Collection.work_on() to a context manager
It allows to open/close objects on start/end of a synchro / work
session. The base method does none of that, but this is a really common
use case. Example: at the beginning of the work session, I open a
webservice client that is available in our WorkContext for the duration
of our work. At the end, I close the client to end the connection.

Example in magento connector:

```
    @contextmanager
    @api.multi
    def work_on(self, model_name, **kwargs):
        self.ensure_one()
        lang = self.default_lang_id
        if lang.code != self.env.context.get('lang'):
            self = self.with_context(lang=lang.code)
        magento_location = MagentoLocation(
            self.location,
            self.username,
            self.password,
            use_custom_api_path=self.use_custom_api_path
        )
        # We create a Magento Client API here, so we can create the
        # client once (lazily on the first use) and propagate it
        # through all the sync session, instead of recreating a client
        # in each backend adapter usage.
        with MagentoAPI(magento_location) as magento_api:
            _super = super(MagentoBackend, self)
            # from the components we'll be able to do: self.work.magento_api
            with _super.work_on(
                    model_name, magento_api=magento_api, **kwargs) as work:
                yield work


```

Commit:
Notify unlink event before the unlink

Otherwise we can't read the record

Commit:
Remove required on the version field

It is no longer necessary for the Components and most of the connectors
redefine it.

lmignon commented:
@guewen What's the maturity of this code? Is-it safe to base new devs on this code?

lmignon committed:
[ADD] setup.py

lmignon committed:
Allow to build components for a specific module on demand.
When a module is loaded with test-enable=True, the components of this module are not registerd by the call to the method _register_hook of the component.builder. Indeed the state of the module is still *to install*. This change allows to call the method *load_components* on the component.builder for a specific module by taking care to not reload the components of a module already loaded.

guewen commented:
@lmignon the magento connector implementation is working with this code (still working on it but the Components part is done) and we are working on a very custom connector based on it. There is a small risk that I need to change a public API if we hit an issue, but nothing too large I presume, it'd more be adjustments (like `Collection.work_on()` that became a context manager). So yes, I encourage to base new devs on this code, which also gives you the opportunity to propose API fine-tunings before it is frozen.

lmignon commented:
@guewen Thank you for your response. You did an awesome job!  I migrated the code from a very specific connector with many components in less than half a day. The API is clean and the code is much cleaner!  The only difficulty encountered was to run the tests, which gave me the opportunity to analyze the code more in depth to fix the problem. (My first small contribution to this big work :smirk:). I agree with you to encourage to base new devs on this code.

Commit:
Simplify tests by loading modules components

It the previous commit, @lmignon added the possibility to load all
components of an addon in a Component Registry.  This commit takes
benefit of this feature to simplify the existing tests and to add a base
test case for the Connector (that loads all components of 'component',
'component_event', 'connector'). It can be used in implementations using
the connector.

guewen commented:
@lmignon happy to read that! And thanks for your contribution, I pushed a new commit that benefits from it already, I could simplify the existing tests and add a ConnectorTransactionCase that loads all the base components in the test `self.comp_registry`.

lmignon commented on connector/tests/common.py:
```diff
@@ -1,47 +1,63 @@
-        yield patched_job
+
+    def setUp(self):
+        super(ConnectorTransactionCase, self).setUp()
```
@guewen `setUpClass` is more appropriate to avoid a reload of the registry for each test.
IMO this changes breaks the modularity. If you write a test for your components you need to know all the modules you depend on that provide components to manually load their components.
You can safely rely on the `_register_hook` method of the `component.builder`model to load all the components of the modules already installed since the method will be called at the end of the registry loading process AND the method clear/clean the registry at each call. Therefore you must only call the `load_components` method for the addon you are testing. Do I miss something?

guewen replied:
@lmignon but `_register_hook` register components in the global component registry for the database, here I want them in the test `comp_registry`. I extracted the content of `_register_hook` in a `build_registry` so we can use it from the setup: 834b351

Indeed, it works well and is more extensible!

Regarding `setUpClass`,  we need to reset the test component registry between each test, not for a whole class. The reason is that if inside a test you add a new component, it should no longer be in the registry for the next test.

```python
    def test_register(self):
        """ Able to register components in components registry """
        class Component1(Component):
            _name = 'component1'

        class Component2(Component):
            _name = 'component2'

        # build the 'final classes' for the components and check that we find
        # them in the components registry
        Component1._build_component(self.comp_registry)
        Component2._build_component(self.comp_registry)
        self.assertEquals(
            ['base', 'component1', 'component2'],
            list(self.comp_registry)
        )

    def test_inherit_bases(self):
        """ Check __bases__ of Component with _inherit """
        class Component1(Component):
            _name = 'component1'

        class Component2(Component):
            _inherit = 'component1'

        class Component3(Component):
            _inherit = 'component1'

        Component1._build_component(self.comp_registry)
        Component2._build_component(self.comp_registry)
        Component3._build_component(self.comp_registry)
        self.assertEquals(
            (Component3,
             Component2,
             Component1,
             self.comp_registry['base']),
            self.comp_registry['component1'].__bases__
        )
```

Commit:
Remove model_name in _event()

It is unused, and depends on the model on which we call it

Cedric-Pigeon commented:
@guewen Fantastic job I based my Lengow connector on your work and I really enjoy the new api. Thanks for that.

Commit:
Add a new method to refine match of components

This method can be used to filter out some components. It is executed on
every candidate component returned by a lookup for a
collection/usage/model so must be kept light to execute if possible.

A use case can be for instance to have more than one kind of export in a
connector for a model.

Say that you have a different way to export customers and suppliers:

    with self.work_on('res.partner', kind='customer') as work:
        exporter = work.component(usage='record.exporter)
        exporter.run()

    with self.work_on('res.partner', kind='supplier') as work:
        exporter = work.component(usage='record.exporter)
        exporter.run()

You can declare your components as:

  class CustomerExporter(Component):

      _name = 'customer.exporter'
      _inherit = ['base.exporter']
      _usage = 'record.exporter'
      _apply_on = ['res.partner']

      @classmethod
      def _component_match(cls, work):
          return work.kind == 'customer'

      def run(self):
          # export customer

  class SupplierExporter(Component):

      _name = 'supplier.exporter'
      _inherit = ['base.exporter']
      _usage = 'record.exporter'
      _apply_on = ['res.partner']

      @classmethod
      def _component_match(cls, work):
          return work.kind == 'supplier'

      def run(self):
          # export supplier

Commit:
Add new TestCase classes to test connectors

When we want to test the implementation of a connector, it must be
simple. Add 2 new test cases for TransactionCase and SavepointCase which
load all the components of the dependencies addons on the tested addons.
Only the components of the addons in state 'installed' are loaded, so we
don't load components of addons that depend on the tested addon.

The tests must call _init_global_registry and build_registry instead of
_register_hook, so the registry is not set to ready. If the global
registry is set to ready during tests, the event will trigger during the
installation of addons which is not what we want.

The existing test case is meant only to be used when one has to
manipulate components in a custom registry.

See discussion on
https://github.com/guewen/connector/commit/447b22f8e36c7e258eba5e8b0d93c1577bd50606#commitcomment-22851711

guewen commented:
@lmignon could you have a look at 1f448f6? Is it more or less what you had in head?

guewen commented on component/builder.py:
```diff
@@ -49,13 +49,20 @@ def _register_hook(self):
-    def build_registry(self, components_registry):
+    def build_registry(self, components_registry, states=None):
+        if not states:
+            states = ('installed', 'to upgrade')
```
I tend to think we don't even need to load the 'to upgrade' components at all?

Commit:
Exclude current addon in ComponentRegistryCase setup

When we run a ComponentRegistryCase test using odoo's way to run tests
(--test-enable), the current addon is in state 'to install' or 'to
upgrade', thus its components not loaded (which is what we want in this
test Case, as it allow us to load manually the components / load
different components). When we run the same test Case from pytest or
nosetest, which are not run during an odoo upgrade, the state of the
addon is 'installed', thus components are loaded.

To ensure the same behavior with --test-enable and external tests
runners, the current addon is always excluded from the components to
load.

Commit:
Set component registry to ready only during tests

Events are not triggered if the component registry is not ready.
Set the component registry to ready and disable it at the end of the
tests, to prevent triggering events during installation of the addon.

Commit:
Fix travis error for missing unittest2

(might be needed in MQT if it works)

Commit:
Add a content index in the migration guide

Commit:
Checkpoint wizard is too slow

Because it calls a name_get per row. As the information is only
repeating the selection, better to hide it.

Commit:
Add readme files

guewen commented:
IMO this branch is now stable enough to be merged :fire: 

Not sure why runbot is red, the log shows 'success', and I'm able to connect on the runbot instance.

We are using it for the Magento connector BTW:
* https://github.com/OCA/connector-ecommerce/pull/41
* https://github.com/OCA/connector-magento/pull/255

Commit:
Bump connector version

Commit:
Fix cache of events

The cache entries for events were including instance of the components,
hence, they included the work instance / odoo environment.

Only the component class and the events it provides are new cached, the
instances of components being instanciated when needed.

Reported on:
https://github.com/OCA/connector-magento/pull/255#issuecomment-315396030

lmignon commented:
@guewen sorry for the delay. Back from holidays I'll review your last changes ASAP...

guewen commented:
> sorry for the delay. Back from holidays I'll review your last changes ASAP...

Really no need to be sorry, holidays well deserved :)

jaredkipe commented:
Inside `core.MetaComponent._build_component` the documentation comments has a hierarchy with two classes `B1` and `B2` that both define `_name = 'b'` but that is not allowed by the check on line 765 that checks if the `_name` is already in the registry.

guewen commented:
> Inside core.MetaComponent._build_component the documentation comments has a hierarchy with two classes B1 and B2 that both define _name = 'b' but that is not allowed by the check on line 765 that checks if the _name is already in the registry.

Good catch! I'm on holidays, will change that when I'm back.

flotho commented on test_component/__manifest__.py:
```diff
@@ -0,0 +1,17 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+{'name': 'Components Tests',
+ 'summary': 'Automated tests for Components, do not install.',
```
Hi @guewen ,

Could you add a description tag so that Travis pass all the build ?

guewen replied:
Done, Travis already passes the build though. Runbot doesn't for an unknown reason but runbot should not block a merge (the runbot instance is up and running BTW).

guewen replied:
Actually this is the reason why runbot was unhappy :) thanks for the hint

Commit:
Allow to add an inheritance on an existing component

Commit:
Make abstract/non-abstract inheritance more robust

* forbid to transform an AbstractComponent to a Component
* forbid to have an AbstractComponent inherit from a Component

Commit:
Add description in test addons to make pylint happier

guewen commented:
@lmignon @lasley @jaredkipe @flotho 
What do you think about merging? Let me know and I'll run a rebase (don't merge before the rebase because I the history has fixup commits).

Once merged, I'll increase the version number, and change the documentation build to use the 10.0 by default (and if time permits, I'll build and push 9.0 and 8.0 docs in subdirectories and add links between versions).

lmignon reviewed (approved):
Great Job!

Cedric-Pigeon reviewed (approved):
Good for me. Great job ! Thanks.

simahawk commented:
:+1:

jaredkipe commented:
Works great for me!

lasley reviewed (approved):
Absolutely amazing @guewen

lmignon commented:
:champagne:

yucer commented:
Hello @guewen,

Which **OCA** connector project do you keep updated to the last version of the connector framework ?

guewen commented:
I can't keep one project updated as it depends on the projects we need for our customers.
Anyway, from the top of my head, the version 10.0 of connector-magento is using the components: https://github.com/OCA/connector-magento/tree/10.0

https://github.com/OCA/connector-jira is as well and is probably the last one I worked on.
